test/e2e/node_feature_discovery.go

```diff
@@ -153,263 +153,285 @@ func cleanupCRDs(cli *nfdclient.Clientset) {
 var _ = SIGDescribe("Node Feature Discovery", func() {
 	f := framework.NewDefaultFramework("node-feature-discovery")
 
-	Context("when deploying a single nfd-master pod", Ordered, func() {
-		var (
-			masterPod *corev1.Pod
-			crds      []*apiextensionsv1.CustomResourceDefinition
-			extClient *extclient.Clientset
-			nfdClient *nfdclient.Clientset
-		)
-
-		BeforeAll(func() {
-			// Create clients for apiextensions and our CRD api
-			extClient = extclient.NewForConfigOrDie(f.ClientConfig())
-			nfdClient = nfdclient.NewForConfigOrDie(f.ClientConfig())
-
-			By("Creating NFD CRDs")
-			var err error
-			crds, err = testutils.CreateNfdCRDs(extClient)
-			Expect(err).NotTo(HaveOccurred())
-		})
-
-		AfterAll(func() {
-			for _, crd := range crds {
-				err := extClient.ApiextensionsV1().CustomResourceDefinitions().Delete(context.TODO(), crd.Name, metav1.DeleteOptions{})
-				Expect(err).NotTo(HaveOccurred())
+	nfdTestSuite := func(useNodeFeatureApi bool) {
+		createPodSpecOpts := func(opts ...testpod.SpecOption) []testpod.SpecOption {
+			if useNodeFeatureApi {
+				return append(opts, testpod.SpecWithContainerExtraArgs("-enable-nodefeature-api"))
 			}
-		})
+			return opts
+		}
 
-		BeforeEach(func() {
-			// Drop the pod security admission label as nfd-worker needs host mounts
-			if _, ok := f.Namespace.Labels[admissionapi.EnforceLevelLabel]; ok {
-				e2elog.Logf("Deleting %s label from the test namespace", admissionapi.EnforceLevelLabel)
-				delete(f.Namespace.Labels, admissionapi.EnforceLevelLabel)
-				_, err := f.ClientSet.CoreV1().Namespaces().Update(context.TODO(), f.Namespace, metav1.UpdateOptions{})
-				Expect(err).NotTo(HaveOccurred())
+		Context("when deploying a single nfd-master pod", Ordered, func() {
+			var (
+				crds      []*apiextensionsv1.CustomResourceDefinition
+				extClient *extclient.Clientset
+				nfdClient *nfdclient.Clientset
+			)
+
+			checkNodeFeatureObject := func(name string) {
+				_, err := nfdClient.NfdV1alpha1().NodeFeatures(f.Namespace.Name).Get(context.TODO(), name, metav1.GetOptions{})
+				if useNodeFeatureApi {
+					By(fmt.Sprintf("Check that NodeFeature object for the node %q was created", name))
+					Expect(err).NotTo(HaveOccurred())
+				} else {
+					By(fmt.Sprintf("Check that NodeFeature object for the node %q hasn't been created", name))
+					Expect(err).To(HaveOccurred())
```

Comment:
IINW, NodeFeature CRD will still be applied even though we are using gRPC because its definition is in the same [file](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/deployment/base/nfd-crds/nfd-api-crds.yaml). I think we should extend the `else` statement to check that NodeFeature is empty.

Reply:
I think that's fine, doesn't do any harm afaics. I mean I think it's ok to create both CRDs even if we run with NodeFeature disabled in the first pass.

Reply:
> My point was, only expecting err to have occurred is not enough, because we might have no error of getting NodeFeature and still get 0 NodeFeature, which is the case that falls under `else{}`.

Ach, sorry for reading this comment way too hastily. This doesn't create the CRD. It's about nfd-worker creating the NodeFeature object (for the node it's running on). We `Get` a NodeFeature object with a specified name in a specified  namespace so we get a NotExist error if that does not exist